Add integration test framework for isolated

andystaples opened the pull request:
- Add integration testing project
- Methods for interacting with "func", "azurite", etc
- One app with basic "Hello Cities" scenario
- Build script that 
  - installs a local copy of Core Tools
  - builds and deploys the extensions to the function app
  - builds the function app
  - installs and launches azurite

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [x] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [ ] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
* [x] My changes **do not** add EventIds to our EventSource logs
    * [ ] Otherwise: Ensure the EventIds are within the supported range in our existing Windows infrastructure. You may validate this with a deployed app's telemetry. You may also extend the range by completing a PR such as [this one](https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/7463263?_a=files).
* [ ] My changes **should** be added to v3.x branch.
    * [ ] Otherwise: This change only applies to Durable Functions v2.x and **will not** be merged to branch v3.x.

andystaples committed:
Add integration test framework for isolated
- One app with basic "Hello Cities" scenario

andystaples committed:
Add license headers, remove unused files

cgillum reviewed (commented):
I suggest a slightly revised file structure for this, basically putting all end-to-end related code under a `test/e2e` directory.

/test/e2e/Apps/DotNetIsolated/App.csproj
/test/e2e/Apps/DotNetIsolated/...
/test/e2e/Tests/E2ETests.csproj
/test/e2e/Tests/...

In the future, we can add other language-specific apps under the /test/e2e/Apps/ directory too.

cgillum commented on test/DotnetIsolatedE2ETests/Fixtures/FixtureHelpers.cs:
```diff
@@ -0,0 +1,89 @@
+    {
+        public static Process GetFuncHostProcess(bool enableAuth = false)
+        {
+            var funcProcess = new Process();
```
nit: let's wait to create the `Process` object until the time when we're actually use it. This is especially important for disposable objects, like `Process`, which can cause memory/handle leaks if they are created but not disposed (which will happen if you create it and we throw an exception in line 25 or line 39).

andystaples replied:
Done

cgillum commented on test/DotnetIsolatedE2ETests/Fixtures/FixtureHelpers.cs:
```diff
@@ -0,0 +1,89 @@
+
+            var e2eAppPath = Path.GetDirectoryName(e2eHostJson);
+
+            var cliPath = Path.Combine(rootDir, @"test/DotnetIsolatedE2ETests/Azure.Functions.Cli/func");
```
This seems to imply that the func.exe tool is checked into the repo, which I assume is not your intention. I recommend putting it into a temporary directory instead.

andystaples replied:
func isn't checked into the repo, but running the build-e2e-test.ps1 script will install it into this location. Useful for several reasons - 
1) Durable "power users" often modify the Core Tools in the PATH, we want an unmodified copy running for these tests
2) Pipeline automation will have to take this step anyway, might as well do it in this script

andystaples replied:
I can modify the install directory - but for the pipeline scenario it is most useful to have it inside the repo folder structure somewhere. Dotnet worker does it in the root.

andystaples replied:
Moved the install location for Core Tools and the start location for azurite to a owned folder in [System.IO.Path]::GetTempPath()/similar.

bachuv commented on test/DotnetIsolatedE2EApps/MainApp/app.csproj:
```diff
@@ -0,0 +1,32 @@
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.21.0" />
+    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.DurableTask" Version="3.1.2" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.DurableTask" Version="1.3.2" />
```
Is it possible to add a project reference to the Worker Extensions package like [this example](https://github.com/Azure/azure-functions-durable-extension/blob/53abf0f5d893340e2ebeb24e352377c5d0c15393/test/SmokeTests/OOProcSmokeTests/DotNetIsolated/DotNetIsolated.csproj#L24) instead of referencing a specific version?

andystaples replied:
Agreed, this would solve several issues with the current solution, including: 
- Dynamic modifying of app.csproj with the build script will lead to unnecessary noise in commits and/or work excluding these files from unrelated PRs
- Would remove the requirement of manually bumping extension versions to run test locally
- Would simplify the build script 

Need to explore this further. The extension build process in isolated has some gotchas but it may be possible to work around

cgillum replied:
It would definitely be better to do a project reference as I mentioned in a separate comment, but we can do that as part of a follow up PR if it's not a simple fix.

andystaples replied:
Done!

bachuv commented on test/DotnetIsolatedE2ETests/Fixtures/FunctionAppFixture.cs:
Can we resolve the warnings that are showing up in this PR?

andystaples replied:
Should be done - let me know if you see specific new ones or ones I missed

bachuv commented on test/DotnetIsolatedE2ETests/DotnetIsolatedE2ETests.csproj:
Can you add a simple README for this project explaining the scripts that we have and how the fixtures and helper files are used?

andystaples replied:
Done

andystaples replied:
I focused more on the build scripts and practical usage steps, happy to add some info on fixtures and helpers if needed but for the most part, the goal is to abstract them to where adding or updating tests will not require interacting with those files at all

bachuv replied:
Thanks!

bachuv reviewed (commented):
Thanks for opening this PR!

andystaples committed:
Implement suggestions from PR review
- Restructure e2e test folder structure
- Add e2e test readme/build instructions
- Move installed temp files to $env:TEMP or equivalent
- Do as many checks as possible before creating Process() object

cgillum commented on test/e2e/Tests/Fixtures/FunctionAppFixture.cs:
```diff
@@ -0,0 +1,165 @@
+                    _jobObjectRegistry.Register(this._funcProcess);
+                }
+
+                var httpClient = new HttpClient();
```
Minor: It's important to dispose `HttpClient` objects when we're doing with them to avoid resource leaks. A simple `using` on this line will ensure that this object gets cleaned up at the end of this method invocation.
```suggestion
                using var httpClient = new HttpClient();
```

cgillum commented on test/e2e/Tests/Helpers/HttpHelpers.cs:
```diff
@@ -0,0 +1,78 @@
+        public static async Task<HttpResponseMessage> InvokeHttpTrigger(string functionName, string queryString = "")
+        {
+            // Basic http request
+            HttpRequestMessage request = GetTestRequest(functionName, queryString);
```
Same suggestion about disposing disposable objects to prevent resource leaks.
```suggestion
            using HttpRequestMessage request = GetTestRequest(functionName, queryString);
```

cgillum commented on test/e2e/Tests/Helpers/HttpHelpers.cs:
```diff
@@ -0,0 +1,78 @@
+        {
+            string uri = $"{Constants.FunctionsHostUrl}/api/{functionName}{queryString}";
+            using (var request = new HttpRequestMessage(HttpMethod.Get, uri))
+            {
```
nit: for this case, you don't need the curly braces with the using statement. Rather, you can save some indenting (which makes the code a bit cleaner) by removing the `{}`. In that case, line 34 would simply be:
```csharp
using var request = new HttpRequestMessage(HttpMethod.Get, uri);
```

andystaples replied:
Done

cgillum commented on test/e2e/Tests/Helpers/HttpHelpers.cs:
```diff
@@ -0,0 +1,78 @@
+                var response = await GetResponseMessage(request);
+
+                Console.WriteLine(
+                    $"InvokeHttpTrigger: {functionName}{queryString} : {response.StatusCode} : {response.ReasonPhrase}");
```
Will console logs be captured in the test output? Normally we write logs in Xunit to `ITestOutputHelper`.

andystaples replied:
Deleting this whole method as it is never used and - IMO - we should rely on the Task<HttpResponseMessage> method and validate responses in tests or methods defined in test classes

cgillum commented on test/e2e/Tests/Helpers/HttpHelpers.cs:
```diff
@@ -0,0 +1,78 @@
+                    string actualMessage = await response.Content.ReadAsStringAsync();
+                    Console.WriteLine(
+                        $"InvokeHttpTrigger: expectedMessage : {expectedMessage}, actualMessage : {actualMessage}");
+                    return actualMessage.Contains(expectedMessage);
```
Could/should we use `Assert.Contains(...)` here instead?

andystaples replied:
Deleting this whole method as it is never used and - IMO - we should rely on the Task<HttpResponseMessage> method and validate responses in tests or methods defined in test classes

cgillum commented on test/e2e/Tests/Fixtures/FixtureHelpers.cs:
```diff
@@ -0,0 +1,79 @@
+using System.Runtime.InteropServices;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Durable.Tests.DotnetIsolatedE2E
```
tip: to reduce indentation, we should use file-scoped namespaces and drop the curly braces. Just need to add a `;` at the end of the `namespace` declaration. I would like us to do this for all new code we write.

```suggestion
namespace Microsoft.Azure.Durable.Tests.DotnetIsolatedE2E;
```

andystaples replied:
Done

cgillum commented on test/e2e/Tests/Tests/HelloCitiesTest.cs:
```diff
@@ -0,0 +1,37 @@
+        [InlineData("HelloCities_HttpStart", "", HttpStatusCode.Accepted, "")]
+        public async Task HttpTriggerTests(string functionName, string queryString, HttpStatusCode expectedStatusCode, string expectedMessage)
+        {
+            HttpResponseMessage response = await HttpHelpers.InvokeHttpTrigger(functionName, queryString);
```
```suggestion
            using HttpResponseMessage response = await HttpHelpers.InvokeHttpTrigger(functionName, queryString);
```

cgillum commented on test/e2e/Tests/Tests/HelloCitiesTest.cs:
```diff
@@ -0,0 +1,37 @@
+            if (!string.IsNullOrEmpty(expectedMessage))
+            {
+                Assert.False(string.IsNullOrEmpty(actualMessage));
+                //Assert.Contains(expectedMessage, actualMessage);
```
Why is this assert commented out?

andystaples replied:
Carryover from the worker test code this code was derived from. Since the response here is a set of API URLs, we don't do any validation yet. Need to write a set of Durable test methods to parse out the StatusQueryGetAPI URL, invoke until orchestration completion, and parse the "actual" function output as a future work item. 
Deleting this line

cgillum commented on test/e2e/Tests/Constants.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public static IConfiguration Configuration = TestUtility.GetTestConfiguration();
+
+        internal static string FunctionsHostUrl = Configuration["FunctionAppUrl"] ?? "http://localhost:7071";
```
nit: since these are in a "Constants" file, we should probably make them readonly.

```suggestion
        internal static readonly string FunctionsHostUrl = Configuration["FunctionAppUrl"] ?? "http://localhost:7071";
```

cgillum commented on test/e2e/Tests/build-e2e-test.ps1:
```diff
@@ -0,0 +1,167 @@
+{
+  Write-Host "---Skipping Core Tools download---"  
+}
+elseif (Test-Path $FUNC_CLI_DIRECTORY) 
```
Seems this `elseif` is redundant with the check above.

andystaples replied:
Removed it

cgillum commented on test/e2e/Tests/e2e-tests-readme.md:
```diff
@@ -0,0 +1,59 @@
+
+## Running the E2E Tests
+
+### Step 1: Increment the host and worker package versions 
```
Is this step necessary if I just want to run the tests without making any changes?

andystaples replied:
Will address with package reference work - this will also involve clearing the local NuGet cache which will make this step redundant

cgillum commented on test/e2e/Apps/BasicDotNetIsolated/app.csproj:
```diff
@@ -0,0 +1,32 @@
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.21.0" />
+    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.DurableTask" Version="3.0.2" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.DurableTask" Version="1.2.2" />
```
I wonder why we wouldn't just use a project reference here instead of a package reference? That way we never have to worry about keeping this version number up to date.

cgillum commented on test/e2e/Apps/BasicDotNetIsolated/app.csproj:
```diff
@@ -0,0 +1,32 @@
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.21.0" />
+    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.DurableTask" Version="3.0.2" />
```
Is there a particular reason to include the WebJobs package reference here? Normally .NET Isolated apps should never reference WebJobs packages directly.

andystaples replied:
Will address along with the work to update to package references

andystaples replied:
Done - now the test app has a reference to the worker extension project only. The webjobs extension is still packed and put in a local nuget packages folder alongside the app, so that the WorkerExtensions.csproj can get the version it needs

cgillum commented on test/e2e/Tests/e2e-tests-readme.md:
```diff
@@ -0,0 +1,59 @@
+    * Building the test app projects
+3. Install and start azurite emulator using Node
+
+NOTE: It should not be necessary to run start-emulators.ps1 manually, as it should be called by the build script. If you have a instance of Azurite already running, it will recognize and skip this step. 
```
Nice

cgillum commented on test/e2e/Tests/e2e-tests-readme.md:
```diff
@@ -0,0 +1,59 @@
+
+## Prerequisites
+
+- Visual Studio IDE
```
Is Visual Studio required or just recommended? Based on what I'm seeing, it doesn't look like we strictly require an IDE to run tests.

andystaples replied:
Removed this line

cgillum commented on test/e2e/Tests/start-emulators.ps1:
```diff
@@ -0,0 +1,198 @@
+#     $IsWindows = $true
+# }
+
+if (!$IsWindows)
```
Consider adding `-and $StartCosmosDBEmulator` so that we don't write this message about skipping if the caller never asked for the cosmos DB emulator.

andystaples replied:
Done

cgillum commented on .gitignore:
```diff
@@ -297,3 +297,9 @@ functions-extensions/
 /src/WebJobs.Extensions.DurableTask/Microsoft.Azure.WebJobs.Extensions.DurableTask.xml
+
+# E2E Tests build output
+/test/DotnetIsolatedE2ETests/Azure.Functions.Cli/*
```
Is this where we put the CLI or does it go into a temp directory (in which case we don't need this .gitignore line)?

andystaples replied:
Removed this and the last line from .gitignore, not needed now

cgillum reviewed (commented):
Looking great. I've added some comments, but I don't think we necessarily need to block this PR on any of them.

andystaples committed:
Apply suggestions from code review

Direct commit applicable PR suggestions

Co-authored-by: Chris Gillum <cgillum@microsoft.com>

andystaples committed:
Further PR improvements
- file-scoped namespaces
- Remove unneeded .gitignore lines
- using statements for HTTP objects where applicable
- small readme improvements

andystaples committed:
Use project reference in e2e test apps

cgillum commented on test/e2e/Tests/build-e2e-test.ps1:
```diff
@@ -111,34 +116,26 @@ Write-Host "Moving nupkg from WebJobs extension to $E2EAppProjectDirectory/packa
+    }
+    
+    # Write-Host "Updating WebJobs.Extensions.DurableTask version to $webJobsExtensionVersion"
+    # dotnet add app.csproj package Microsoft.Azure.WebJobs.Extensions.DurableTask --version $webJobsExtensionVersion
```
Can these commented-out lines be removed?

cgillum reviewed (approved):
LGTM! As a follow up to this PR, let's create another PR to run these tests as part of the CI.

andystaples committed:
Remove comments
